### PR TITLE
Fix codegen for `__apply` functions

### PR DESCRIPTION
**Pull request by aq17**
# Description

Fixes `pulumi/yaml` [337](https://github.com/pulumi/pulumi-yaml/issues/337)

minor: Fixes https://github.com/pulumi/pulumi/issues/8077

## Checklist

- [x] I have added tests that prove my fix is effective or that my feature works

- [x] I have run `make changelog` and committed the `changelog/pending/<file>` documenting my change

- [ ] Yes, there are changes in this PR that warrants bumping the Pulumi Service API version

### CODE REVIEW

File: pkg/codegen/go/gen_program_expressions.go [hunk 1]
```diff
@@ -96,7 +96,13 @@ func (g *generator) genAnonymousFunctionExpression(
 	body, temps := g.lowerExpression(expr.Body, retType)
 	g.genTempsMultiReturn(w, temps, retTypeName)
 
-	g.Fgenf(w, "return %v, nil", body)
+	// fromBase64 special case
+	if strings.Contains(fmt.Sprintf("%v", body), "fromBase64") {
```
Comment (iwahbe) [inline pkg/codegen/go/gen_program_expressions.go:100]: This seems like a very fragile check. Is there a different way we can do this?

File: pkg/codegen/go/gen_program_expressions.go [hunk 2]
```diff
@@ -213,8 +219,6 @@ func (g *generator) GenFunctionCallExpression(w io.Writer, expr *model.FunctionC
 		default:
 			g.Fgenf(w, "%.v", expr.Args[0])
 		}
-	case pcl.IntrinsicApply:
```
Comment (iwahbe) [inline pkg/codegen/go/gen_program_expressions.go:216]: Can you explain why we are removing this?

File: pkg/codegen/nodejs/gen_program_expressions.go [hunk 3]
```diff
@@ -435,6 +435,8 @@ func (g *generator) GenFunctionCallExpression(w io.Writer, expr *model.FunctionC
 		g.Fgenf(w, "%.20v.split(%v)", expr.Args[1], expr.Args[0])
 	case "toBase64":
 		g.Fgenf(w, "Buffer.from(%v).toString(\"base64\")", expr.Args[0])
+	case "fromBase64":
+		g.Fgenf(w, "Buffer.from(%v).toString(\"ascii\")", expr.Args[0])
```
Comment (iwahbe) [inline pkg/codegen/nodejs/gen_program_expressions.go:439]: I think this should be some form of unicode, not ascii.

File: pkg/codegen/testing/test/testdata/functions-pp/dotnet/functions.cs [hunk 4]
```diff
@@ -6,7 +6,9 @@
 {
     var encoded = Convert.ToBase64String(System.Text.Encoding.UTF8.GetBytes("haha business"));
 
-    var joined = string.Join("-", new[]
+	var decoded = Encoding.UTF8.GetString(Convert.FromBase64String(encoded));
+
+	var joined = string.Join("-", new[]
```
Comment (iwahbe) [inline pkg/codegen/testing/test/testdata/functions-pp/dotnet/functions.cs:11]: The indentation is wrong here. 
```suggestion
    var decoded = Encoding.UTF8.GetString(Convert.FromBase64String(encoded));

    var joined = string.Join("-", new[]
```

File: pkg/codegen/go/gen_program_expressions.go (hunk 1 above)
Comment (aq17) [reply pkg/codegen/go/gen_program_expressions.go:100]: Unfortunately as an `AnonymousFunctionExpression`, we don't have an explicit name.  `fromBase64` is our internal function mapping used to prefix the body (i.e. `fromBase64(__convert(aksCredentials.kubeconfigs[0].value))` – would [centralizing all of these](https://github.com/pulumi/pulumi/blob/master/pkg/codegen/go/gen_program_expressions.go#L218-L336) into one place suffice?

Comment (iwahbe) [reply pkg/codegen/go/gen_program_expressions.go:100]: Somehow, `body` knows how to format itself to include the string `fromBase64`. Do you know where this string is stored in `body`?

Comment (aq17) [reply pkg/codegen/go/gen_program_expressions.go:100]: It should be defined [here](https://github.com/pulumi/pulumi/blob/aqiu/10726/pkg/codegen/pcl/functions.go#L296) in `functions.go`, referenced by the `[binder](https://github.com/pulumi/pulumi/blob/master/pkg/codegen/pcl/binder.go#L133-L135)`

**Comment by aq17** [comment]
@iwahbe I modified `functions.pp` so it will generated compilable Go code

File: pkg/codegen/go/gen_program_expressions.go (hunk 1 above)
Comment (iwahbe) [reply pkg/codegen/go/gen_program_expressions.go:100]: After https://github.com/pulumi/pulumi/pull/10775/commits/ea4fd10299604846b9bdfef20f64207bd6b9c1f3 landed, I ran the tests with some added debug printing. It's not a `AnonymousFunctionExpression`. Its a `FunctionCallExpression`, which makes sense. Testing on my machine, this works:
```suggestion
	if body, ok := body.(*model.FunctionCallExpression); ok && body.Name == "fromBase64" {
```

File: pkg/codegen/go/gen_program.go [hunk 5]
```diff
@@ -808,13 +809,17 @@ func (g *generator) genLocalVariable(w io.Writer, v *pcl.LocalVariable) {
 			assignment = "="
 		}
 	}
+
```
Comment (iwahbe) [inline pkg/codegen/go/gen_program.go:812]: Is it possible to hoist this outside of the switch. It's called that everywhere except for the `!noError` case in `pcl.Invoke` and in the `default:` case. It should be called in the `default:` case.
```suggestion
	if name == "_" {
 		assignment = "="
 	}
```

File: pkg/codegen/go/gen_program.go [hunk 6]
```diff
@@ -823,10 +828,30 @@ func (g *generator) genLocalVariable(w io.Writer, v *pcl.LocalVariable) {
...
+			"toBase64":
+			if name == "_" {
+				assignment = "="
+			}
+			g.Fgenf(w, "%s %s %.3v;\n", name, assignment, expr)
+		case "fromBase64":
+			tmpVar := fmt.Sprintf("%s%d", "tmpVar", g.tmpVarCount)
+			g.Fgenf(w, "%s, _ %s %.3v;\n", tmpVar, assignment, expr)
+			if name == "_" {
+				assignment = "="
```
Comment (iwahbe) [inline pkg/codegen/go/gen_program.go:850]:
```suggestion
				assignment = ":="
```

**Review by iwahbe — APPROVED** [review]
I noticed a couple last things to fix. They are very small. Otherwise LGTM!

**Comment by aq17** [comment]
bors merge

**Pull request merged**